backupccl: allocate work on index level

msbutler opened the pull request:
Previously, backup planning would pass a list of merged index spans for
PartitionSpans to assign to nodes. With this patch, a list of unmerged
index spans are now passed to PartitionSpans.

Previously, online restore could create a virtual sst that pointed to empty key
space which would then cause pebble to panic during the download phase. This
was possible because of the following:

1.  backup could previously create a backup file that spanned a table key
space, but which contained an empty index.
2. restore allocates work using unmerged table index spans
3. online restore creates virtual sst spans by intersecting each backup file
with each restore span entry. The virtual sst for this empty index would then
point to empty key space.

This patch prevents this empty index span from being written to a backup file,
thus preventing this online restore vulnerability.

Informs https://github.com/cockroachdb/cockroach/issues/122176

Release note: none

msbutler committed:
backupccl: allocate work on index level

Previously, backup planning would pass a list of merged index spans for
PartitionSpans to assign to nodes. With this patch, a list of unmerged
index spans are now passed to PartitionSpans.

Previously, online restore could create a virtual sst that pointed to empty key
space which would then cause pebble to panic during the download phase. This
was possible because of the following:

1.  backup could previously create a backup file that spanned a table key
space, but which contained an empty index.
2. restore allocats work using unmerged table index spans
3. online restore creates virtual sst spans by intersecting each backup file
with each restore span entry. The virtual sst for this empty index would then
point to empty key space.

This patch prevents this empty index span from being written to a backup file,
thus preventing this online restore vulnerability.

Informs #122176

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/122471)
<!-- Reviewable:end -->

msbutler committed:
sql: add PartitionSpansWithoutMerging variant

PartitionSpans automatically merges adjacent spans assigned to the same node.
This patch introduces a variant of PartitionSpans which does not do this.

Epic: none

Release note: none

msbutler committed:
backupccl: unskip online restore variant of TestBackupRestoreRandomRoundtrips

Epic: none

Release note: none

dt commented:
Do we still need range-sized-spans splitting in backup processor with this? doesn't PartitionSpans already split the spans up with an RDI, and with `WithoutMerging` it won't stitch them back up, right?

msbutler commented:
I was planning to remove range-sized-spans splitting logic after a few nightly runs with this patch.

Also removed all "table-index" references in the commit message.

msbutler commented:
TFTR! 

bors r=dt
